src/main/java/de/tum/in/www1/artemis/service/scheduled/SchedulingService.java

```diff
@@ -0,0 +1,67 @@
+package de.tum.in.www1.artemis.service.scheduled;
+
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Profile;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.scheduling.support.CronExpression;
+import org.springframework.stereotype.Service;
+
+import de.tum.in.www1.artemis.service.plagiarism.ContinuousPlagiarismControlService;
+
+@Service
+@Profile("scheduling")
+public class SchedulingService {
+
+    private static final Logger log = LoggerFactory.getLogger(SchedulingService.class);
+
+    /**
+     * Daily triggers plagiarism checks as a part of continuous plagiarism control.
+     */
+    @Value("${artemis.continuous-integration.image-cleanup.cleanup-schedule-time:0 0 3 * * *})")
+    private String continuousPlagiarismCleanUpCron; // Example property loaded from application.yml
+
+    /**
+     * Every minute, query for modified results and schedule a task to update the participant scores.
+     * We schedule all results that were created/updated since the last run of the cron job.
+     * Additionally, we schedule all participant scores that are outdated/invalid.
+     */
+    private final String participantScoreCron = "0 * * * * *";
+
+    private final List<TaskConfiguration> tasks = new ArrayList<>();
+
+    public SchedulingService(ContinuousPlagiarismControlService continuousPlagiarismControlService, ParticipantScoreScheduleService participantScoreScheduleService) {
+        // Example scheduled interval task - runs every 30 seconds
+        tasks.add(new TaskConfiguration("IntervalTask", Duration.ofSeconds(30), () -> log.info("Running Interval Task"), LocalDateTime.now()));
+
+        // Examples for tasks with cron expressions
+        tasks.add(new TaskConfiguration("ContinuousPlagiarismCleanupTask", continuousPlagiarismCleanUpCron, continuousPlagiarismControlService::executeChecks));
+        tasks.add(new TaskConfiguration("ParticipantScoreTask", participantScoreCron, participantScoreScheduleService::scheduleTasks));
+    }
+
+    @Scheduled(fixedRate = 1000)
+    public void scheduleTasks() {
```

Comment:
I’m not sure about this approach:

The native Spring scheduling will put the threads to sleep. For this scenario the existing schedulers in the Java runtime and operating systems have efficient implementations to wake up the thread only when really necessary.
With this new approach we would be checking every second if a task needs to be executed. Since we have a wide range of time between task executions (once a minute for the score updater, once a day for most, once a week for token update), we would even in the base case unnecessarily execute this check $\frac{59}{60} \approx 98.3\\%$ of the time., and in $\frac{86399}{86400}$ cases for the most frequent intervals.
I concede that every execution only does little work, but I still think the busy waiting should be avoided.

---

Two alternative approaches that I think would be better:

- Ensure with an architecture test that all methods annotated with `@Scheduled` are only declared in classes residing in the `artemis.service.scheduled` package.
- Use this `SchedulingService` class as a central location where multiple methods exist that are annotated each with their respective `@Scheduled` and then only forward the call to the respective service.
  - If we want to make sure the scheduled tasks do not interrupt any test behaviour, we could simply exclude this class here from the `test` profile.
    - The behaviour of the actual implementation could still be tested by calling the actual service method directly in a test.
    - The `ScheduleService` could be tested fully independent of any Spring application start by manually injecting the relevant Mocks in the constructor and ensuring the relevant methods are called similar to how the `VcsTokenRenewalServiceTest` is implemented.